src/upgrades/mod.rs

```diff
@@ -129,94 +260,175 @@ pub trait UpgradeStore {
     fn event_key(scope: &str, nr: u64) -> KeyStoreKey {
         KeyStoreKey::scoped(scope.to_string(), format!("delta-{}.json", nr))
     }
+}
 
-    fn archive_snapshots(&self, scope: &str) -> Result<(), UpgradeError> {
-        let snapshot_key = KeyStoreKey::scoped(scope.to_string(), "snapshot.json".to_string());
-        let snapshot_bk_key = KeyStoreKey::scoped(scope.to_string(), "snapshot-bk.json".to_string());
-
-        if self.store().has(&snapshot_key)? {
-            self.archive_to_migration_scope(&snapshot_key)?;
-        }
-
-        if self.store().has(&snapshot_bk_key)? {
-            self.archive_to_migration_scope(&snapshot_bk_key)?;
+/// Prepares a Krill upgrade related data migration. If no data migration is needed
+/// then this will simply be a no-op. Returns the [`KrillUpgradeVersions`] if the currently
+/// deployed Krill version differs from the code version. Note that the version may
+/// have increased even if there is no data migration needed.
+///
+/// In case data does not to be migrated, then the this data will be prepared under
+/// the directory returned by `config.data_dir()`. By design, this migration can be
+/// executed while Krill is running as it does not affect any current state. If can
+/// be called multiple times and it will resume the migration from the point it got
```

Comment:
It can be called multiple times but not concurrently, right?

Reply:
right - the lockfile still needs to be done before closing this. Leaving this open as a reminder

Reply:
I added lockfile support in the latest commit. We use similar code in krill-sync.